wtransport/src/config.rs

```diff
@@ -29,6 +29,11 @@ impl ServerConfig {
     pub fn builder() -> ServerConfigBuilder<WantsBindAddress> {
         ServerConfigBuilder::default()
     }
+
+    /// Tweak the given [`rustls::ServerConfig`] to the needs of [`wtransport`].
+    pub fn tweak_tls_config(tls_config: &mut rustls::ServerConfig) {
```

Comment:
Why do we need this?

We can add the ALPN afterwards, when the config is passed within `with_tls_config`.

Reply:
This is exposed so that any consumer that desires to provide their own `rustls::ServerConfig` can apply the tweaks that are required by `wtransport`.

While we can add them, what if the consumer of the API does not want that?

There are definitely other options to do that, like maybe exposing two methods - one that would directly set the `tls_config` value, and one that would set APLN and then set the `tls_config` value.

Or, maybe we can omit the one that doesn't set the APLN altogether - since we have an even more generic one that takes a quinn config. But then, we get back to the idea that this call is a nice-to-have because users can call it on their `rustls::ServerConfig` at any time and build the quinn config from that. So, that was my general idea behind this API.

It is like, what would I really like to have at my disposal if I were to build my own `rustls` config? This thing.

Reply:
This forces a mutable state by splitting state configuration. Moreover, I think it is quite hard to understand what the method really does from the user point of view.

I am tempted to say that if the one is injecting the TLS configuration he/she knows what is dealing with here. WebTransport ALPN (that is, actually HTTP3 ALPN) is already expose in `wtransport-proto`.

I would rather leave the flexibility to the user to pass whatever TLS configuration they want where we blindly accept it

Reply:
> This forces a mutable state by splitting state configuration.

I don't get it. Could you elaborate please?

It's just a static call, it is supposed to be used like `wtransport::ServerConfig::tweak_tls_config(&mut my_tls_config)` before the said config is passed wherever down the line. Of course, it requires a mutable state, but `&mut` is actually the weakest access possible here since we are altering the config field. Taking the config by value and returning it would be worse here, cause sometimes you don't own the value by can mutate it - if that's what you mean.

Reply:
It is not like this call mutates the `wtransport::ServerConfig`. It could be a standalone fn in the module, but I figured it makes sense to group it like this, because the client would have a similar method. It makes sense if you think how this API would be called: users will most likely have to import the `ServerConfig`, and it will contain this handy method for them if they need to apply the tweaks.

This (both the intent and the recommended way to build the `wtransport::ServerConfig`) should of course be properly documented to avoid confusion.

Reply:
> I don't get it. Could you elaborate please?

This force the user of the API having a mutable state on the TLS configuration.

For example:
```
let mut my_tls_config = build_custom_tls_config();
      ^------ I need this mut
// wtransport
ServerConfig::tweak_tls_config(&mut my_tls_config);

// use `my_tls_config` afterwards
```

Moreover it introduces a hidden logical API depdendency: "you should* call `ServerConfig::tweak_tls_config` _before_ passing your tls config to wtransport".

*should: the user can skip the auto-set of ALPN; but generally this is a smell, something is wrong. By standard, WebTransport (actually HTTP3) uses that specific ALPN. Something else might still work, but it is altering the nature of the protocol.

Logical code dependency is bad, and Rust typing system generally helps with that. That's why `ServerConfig` and `ClientConfig` are structured with that design in `wtransport`. User cannot create a bad configuration (e.g. missing to call some method) simply because the type system does not allow that at compile time.

Dependency injection (the purpose of this CR) is very flexible (user can pass whaterver TLS configuration is needed) but more error prone. Possibly, a TLS configuration not valid with the protocol itself (e.g., ALPN different than what is expected). 

For now, as we already agreed, we can use dependency injection but I would suggest to limit the scope (simply because it's a pattern which is not really congruent with the design of wtransport API, as mentioned before).

Regarding ALPN, I personally would remove this method from the public interface.

Reply:
> Moreover it introduces a hidden logical API depdendency: "you should* call `ServerConfig::tweak_tls_config` _before_ passing your tls config to wtransport".

What is the alternative though? Make it so users don't have to call this method?

If the users use a builder pattern as usual they wouldn't need it, if they use a low-level interface - they'd need to set the value, unless we want to force-overwrite it. Do we?

> *should: the user can skip the auto-set of ALPN; but generally this is a smell, something is wrong. By standard, WebTransport (actually HTTP3) uses that specific ALPN. Something else might still work, but it is altering the nature of the protocol.

I agree! Maybe we should address my concerns here separately, which are: how can we reuse the same quinn connector with both WebTransport and other protocols.
I have a use case where it would make sense to serve the classic HTTP3 pages from the same we socket as WebTransport, but also an elaborate custom proxy-protocol. Would be great if that could also be supported. The second case would require tweaking APLN in a way that adds the WebTransport instead of overwriting it, so it doesn't make that much sense. It would make more sense if we made this fn to add the APLN to the list instead of just assiging it...



> Logical code dependency is bad, and Rust typing system generally helps with that. That's why `ServerConfig` and `ClientConfig` are structured with that design in `wtransport`. User cannot create a bad configuration (e.g. missing to call some method) simply because the type system does not allow that at compile time.

Yes, I get that, let's try to figure out a way to do this in a way that doesn't allow invalid states. The question is what states should we allow... I'd argue that an empty APLNs list is a valid state *if* the user uses the low-level interface *and* explicitly communicates that the standard WebTransport APLN is undesirable or a custom one is present. This makes for good defaults even for a low-level interface (i.e. set APLN unless told not to) and still enables the option to not set it if needed.

Reply:
I agree. I have created a zulip server for this crate; the link is in the README. Maybe we can open a thread there and explore solutions.

In the meanwhile, regarding this PR, what if we just remove `tweak_tls_config` and completely rely on the customer configuration? We could also perform a check during the wtransport config build and append the HTTP3 ALPN if not already there. However, to me, it also works if we just rely on user (it's a low level api after all)